JS Components: Add spacing props to Text component

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR implements a spacing feature for the `<Text />` component.

The new props supported allow define the box-model for the component (margin and padding, not border)
Prop | Desc
------|------
`m` |  margin
`mt` | margin-top
`mr` | margin-right
`mb` | margin-bottom
`ml` | margin-left
`mx` | margin left and right
`my` | margin top and bottom
`p` |  padding
`pt` | padding-top
`pr` | padding-right
`pb` | padding-bottom
`pl` | padding-left
`px` | padding left and right
`py` | padding top and bottom

Example:
```jsx
function MyBoxModel() {
  // margin-left: 24px
  // margin-right: 24px
  // padding-bottom: 48px
  return (
    <Text mx={ 3 } pb={ 6 }>Hello</Text>
  );
}
```

Fixes #

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* JS Components: Add spacing props to Text component

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to My Jetpack overview
* Confirm the title of Plans and Connection sections are correct
  * font-size: 36px
  * margin-bottom: 24px;
<img width="431" alt="Screen Shot 2022-03-29 at 3 09 50 PM" src="https://user-images.githubusercontent.com/77539/160677310-0b0b702c-b077-402e-ae8c-91b19f1d10fa.png">

* Using Storybook
* Select the Box Model story of the Text component
* Check the spacing properties

https://user-images.githubusercontent.com/77539/160678313-b3b82fde-fc7c-472d-86ed-380402163f8e.mov

retrofox committed:
[not verified] js-components: introduce Text spacing props

retrofox committed:
[not verified] js-components: set heading margins using spacing props

retrofox committed:
[not verified] js-components: rename spacing with mt, mr, mb and ml

retrofox committed:
[not verified] js-components: add pt, pr, pb and pl props

retrofox committed:
[not verified] js-components: add m and o prosp to Text

retrofox committed:
[not verified] js-componets: add mx, my, px, py props to Text

retrofox committed:
[not verified] js-components: allow set zero to Text box-model props

retrofox committed:
[not verified] js-components: improve generating Text CSS classes

retrofox committed:
[not verified] js-components: add box-model story to Text

retrofox committed:
[not verified] js-components: extends spacing to 8 units

retrofox committed:
[not verified] changelog

renatoagds commented on projects/js-packages/components/components/text/index.jsx:
```diff
@@ -32,11 +32,41 @@ export const TIPOGRAPHY_SIZES = {
+	// Build Styles module CSS classnames.
+	const boxModelClasses = BOX_MODEL_VALUES.reduce( ( acc, value ) => {
+		if ( typeof componentProps[ value ] !== 'undefined' ) {
+			acc += styles[ `${ value }-${ componentProps[ value ] }` ] + ' ';
```
We need to get directly these props and not use them from `componentProps`, which is used for composition and normally applies directly to first-class tags `<h1>, <p>`, etc ... 

This way we'll probably receive some warnings

retrofox replied:
sorry I don't get it. `componentProps` is _the rest of the component props_, right? In the end, it pulls the props from the function argument.

renatoagds replied:
@retrofox The rest props are designed to pass down props when you're using the `component` prop. The `component` props are designed to be used for the cases that you want the `text` styles and not use its default `component` for that moment.

Example of this usage:

https://github.com/Automattic/jetpack/blob/master/projects/packages/my-jetpack/_inc/components/product-detail-card/index.jsx#L218-L226

In this case, we're using the `Text` component only to get the `typography` styles, but the final component that renders is `ProductDetailButton`, and we pass some props that are for `ProductDetailButton` and not `Text`.

That's the reason we get what we need at `Text` and pass down only the rest.

The other reason is that `Unknown Prop Warning`: https://reactjs.org/warnings/unknown-prop.html 

Is a warning related to passing non-standard props for `DOM` elements, and we're mostly based on `DOM` elements.

retrofox replied:
I think I get that. The idea is picking those props that in some way belong to the `<text />` component. It means typography but also now spacing props. The rest of props should be pass down to the component:

```jsx
<Text
	component={ ProductDetailButton }
	onClick={ clickHandler }
	isLoading={ isFetching }
	isPrimary={ ! isBundle }
	href={ onClick ? undefined : addProductUrl }
	className={ styles[ 'checkout-button' ] }
	variant="body"
>
	Click me!
</Text>
``` 

Clearly, above, the props that belong to the Text component are `variant` and `component`, correct? We could do something like...

```jsx
<Text
	component={ ProductDetailButton }
	onClick={ clickHandler }
	href={ onClick ? undefined : addProductUrl }
	className={ styles[ 'checkout-button' ] }
	variant="body"
	mx={ 3 }
	pb={ 3 }
>
	Click me!
</Text>
```

Here, `component`, `variant`, `mx` and `pb` belong to Text component, and won't be propagated to the component instance.

retrofox replied:
I'm deleting the spacing props in the line below:
```jsx
delete componentProps[ value ];
```

renatoagds replied:
Oh! Missed the `delete` part, sorry. 🚀

retrofox replied:
yeah, it deserves a comment tbh :yesyes:

retrofox replied:
https://github.com/Automattic/jetpack/pull/23682/commits/9f8b111cb0c488e40c33e15053a583e7d9cc7fd4

renatoagds commented on projects/js-packages/components/components/text/stories/index.jsx:
```diff
@@ -6,11 +6,56 @@ import React from 'react';
+		py: {
+			control: { type: 'select', options: SPACING_VALUES },
+		},
+	},
```
You could simplify this one exporting `BOX_MODEL_VALUES`:

```es6
argTypes: {
  ...BOX_MODEL_VALUES.reduce((obj, cur) => ({
    ...obj, [cur]: {
      control: { type: 'select', options: SPACING_VALUES },
    }
  }), {}),
}
```

retrofox replied:
https://github.com/Automattic/jetpack/pull/23682/commits/a57ef95b3bdf5e4c6042d236123493a79dca0078

renatoagds commented on projects/js-packages/components/components/text/style.module.scss:
```diff
@@ -70,14 +70,45 @@
+);
+
+@each $prop, $p in $props {
+	$sizes: ( 0, 1, 2, 3, 4, 5, 6, 7, 8 );
```
This put a limit on us, that's the reason I vote for us to move for some `css-in-js` solution that helps with this integration between CSS and JS, but let's keep this for the future.

retrofox replied:
agree, the style and js are not connect which means we cannot share data between them.

renatoagds replied:
yes, at my P2: p9dueE-3Tq-p2 I highlighted this but waited to get more use cases to make this movement clearer.

retrofox committed:
[not verified] js-components: fix margin-bottom for H3

retrofox committed:
[not verified] js-components: improces Text story
props to @renatoagds https://github.com/Automattic/jetpack/pull/23682#discussion_r837819175

retrofox committed:
[not verified] js-components: add comment

renatoagds reviewed (dismissed):
LGTM

retrofox committed:
bump jetpack-components version

keoshi commented:
Cool stuff!

Does `var( --spacing-base ) === 8px`?

retrofox commented:
> Cool stuff!
> 
> Does `var( --spacing-base ) === 8px`?

Exactly. All text components are composed via `<Text />` component, and spacing wise, the unit is defined by `--spacing-base` which is 8px.

For instance, for Plans section and Status connection, the title is rendered by the `<H3 />` component:
markup | box-model
------|------
<img src="https://user-images.githubusercontent.com/77539/160822693-07cd30bd-6885-4ab8-b30d-c32db185813a.png" width="400px" /> | <img width="196" alt="image" src="https://user-images.githubusercontent.com/77539/160822803-d4099f90-e4f0-4984-9b48-bbaad400f49a.png">

`<H3 />` simply creates a `<Text />` component that defines the typography and spacing:

```jsx
function H3( { children }) {
  return (
    <Text variant="headline-small" mb={ 3 }>
      { children }
    </Text>
  );
}
````

`mb={ 3 }` sets `margin-bottom: 24px` => `3 x 8px`

retrofox commented:
> Great news! One last step: head over to your WordPress.com diff, D77784-code, and deploy it. Once you've done so, come back to this PR and add a comment with your changeset ID.

r242774-wpcom
